### ISSUE TITLE
Вопросы по роутам

### ISSUE THREAD

**Issue by boxfrommars** [Rutorika/laravel-rules#1]
#### resource-роуты

существует метод для регистрации роутов для restful-контроллеров:

``` php
Route::resource('photo', 'PhotoController');
```

генерирует следующие соответствия:

| Verb | Path | Action | Route Name |
| --- | --- | --- | --- |
| GET | `/photo` | index | photo.index |
| GET | `/photo/create` | create | photo.create |
| POST | `/photo` | store | photo.store |
| GET | `/photo/{photo}` | show | photo.show |
| GET | `/photo/{photo}/edit` | edit | photo.edit |
| PUT/PATCH | `/photo/{photo}` | update | photo.update |
| DELETE | `/photo/{photo}` | destroy | photo.destroy |

также можно генерировать только необходимые роуты:

``` php
Route::resource('photo', 'PhotoController', ['only' => ['index', 'show']]);
// or
Route::resource('photo', 'PhotoController', ['except' => ['create', 'store', 'update', 'destroy']]);
```

**плюсы:**
- стандартное решение
- совместимость с rest

**минусы:**
- обычно не нужна такая детализация методов, как при дефолтной генерации (например, `edit`, `create` и `show` обычно одна и та же страница)
#### controller-роуты

также существует [способ описания controller-роутов](http://laravel.com/docs/5.1/controllers#implicit-controllers), то есть 

``` php
Route::controller('users', 'UserController');
```

а дальше каждый метод контроллера вида `getSomething`, `postSomething` etc автоматически регистрирует соответственный роут (`GET /user/something` и `POST /user/something`)

**плюсы**
- быстрое добавление методов-роутов, не нужно дополнительно ничего прописывать

**минусы**
- плохая читаемость, чтобы увидеть все роуты контроллера необходимо прочитать его весь (или выполнить `php artisan route:list`)
- контроллер занимается не своей ответственностью (описывает урлы приложения)
- невозможно использовать один метод для разных урлов с разными параметрами
### ограничения на параметры

[laravel позволяет накладывать ограничения на параметры](http://laravel.com/docs/5.1/routing#route-parameters) в виде регулярных выражений:
локально:

``` php
Route::get('user/{id}', /* ... */})->where('id', '[0-9]+');
```

или глобально (в `RouterServiceProvider`)

``` php
$router->pattern('id', '[0-9]+');
```

будем ли мы использовать ограничения? 

**плюсы:**
- можно не беспокоиться о виде параметра в контроллере

**минусы:**
- избыточный код. обычно мы используем в урлах идентификаторы, поэтому неважно, что нам пришло, мы всё равно в контроллере просто вызываем ->find($parameter) и обрабатываем ненахождение 404 ошибкой. т.к. используется PDO абсолютно безопасно использовать любой вид параметра
### Model binding

статус [этой возможности](http://laravel.com/docs/5.0/routing#route-model-binding) в laravel на данный момент не известен, существуют опасения, что этот функционал будет признан deprecated, т.к. его описание исключено из документации к 5.1 (ссылка выше дана на документацию к 5.0), обсуждение см. в 
- https://github.com/laravel/docs/commit/00be759cea21d6b4d59c6ce3814042599e5223f4#diff-ce107015c43d323e34bbf68cbc83e623 
- https://github.com/laravel/framework/issues/9403 
- https://github.com/laravel/framework/issues/9381

поэтому не думаю, что пока стоит обсуждать использование. я постараюсь следить за статусом и как только что-то прояснится, вынесу на обсуждение

**Comment by boxfrommars** [comment]
примеры роут-файлов в сторонних проектах
- https://github.com/sevenshores/kraken/blob/master/app/Http/routes.php
- https://github.com/AbuseIO/AbuseIO/blob/master/app/Http/routes.php
- https://github.com/bestmomo/laravel5-example/blob/master/app/Http/routes.php
- https://github.com/laravelio/laravel.io/blob/master/app/Http/routes.php
- https://github.com/rappasoft/laravel-5-boilerplate/blob/master/app/Http/routes.php (с разделением на файлы)

**Comment by boxfrommars** [comment]
примеры в наших проектах

#### https://bitbucket.org/rutorika/compasspools/src/562987012b52e2cf9f550a0737adb720721944c4/php/app/routes.php?at=master&fileviewer=file-view-default

``` php
$regId = '[0-9]+';
$regUrl = '[a-zA-Z0-9_-]{1,250}';
// ...
$entities = array(
            'users'           => 'UsersController',
            'pools'           => 'PoolsController',
            'pool-types'      => 'PoolTypesController',
            'dealers'         => 'DealersController',
            'dealer-contacts' => 'DealerContactsController',
            'galleries'       => 'GalleriesController',
            'cities'          => 'CitiesController',
        );

        foreach ($entities as $entity => $controller) {
            Route::get( "/{$entity}",              array("as" => "admin.{$entity}",          "uses" => "{$controller}@index"));
            Route::get( "/{$entity}/view",         array("as" => "admin.{$entity}.view-new", "uses" => "{$controller}@view"));
            Route::get( "/{$entity}/{id}/view",    array("as" => "admin.{$entity}.view",     "uses" => "{$controller}@view"))->where("id", $regId);
            Route::post("/{$entity}/save",         array("as" => "admin.{$entity}.save-new", "uses" => "{$controller}@save"));
            Route::post("/{$entity}/{id}/save",    array("as" => "admin.{$entity}.save",     "uses" => "{$controller}@save"))->where("id", $regId);
            Route::any( "/{$entity}/{id}/publish", array("as" => "admin.{$entity}.publish",  "uses" => "{$controller}@publish"))->where("id", $regId);
            Route::any( "/{$entity}/{id}/delete",  array("as" => "admin.{$entity}.delete",   "uses" => "{$controller}@delete"))->where("id", $regId);
        }
```

за исключением того, что данный подход добавляет метод `publish` в основном он нормально преобразуется в resource подход. в свою очередь publish реализуется с помощью `PATCH` запроса.

#### https://bitbucket.org/ustaxi/fasten_site/src/a930f4e9af2eeac0b089798a8a9ca37b34ea4e61/app/Http/routes.php?at=master&fileviewer=file-view-default

``` php
Route::group(['prefix' => 'cities'], function () {
            Route::get('/', "\\Admin\\Controllers\\ContactsController@viewList");
            Route::get('/sync', "\\Admin\\Controllers\\ContactsController@sync");
            Route::get('/{id}', "\\Admin\\Controllers\\ContactsController@viewCity");
            Route::post('/{id}/update', "\\Admin\\Controllers\\ContactsController@updateCity");
            Route::post('/{id}/hide', "\\Admin\\Controllers\\ContactsController@hide");
            Route::post('/{id}/show', "\\Admin\\Controllers\\ContactsController@show");
        });

        Route::group(['prefix' => 'flat-rates'], function () {
            Route::post('/create', "\\Admin\\Controllers\\FlatRatesController@create");
            Route::post('/update/{id}', "\\Admin\\Controllers\\FlatRatesController@update");
            Route::post('/delete/{id}', "\\Admin\\Controllers\\FlatRatesController@delete");
        });

        Route::group(['prefix' => 'legals'], function () {
            Route::get('/', "\\Admin\\Controllers\\LegalsController@viewList");
            Route::get('/create', "\\Admin\\Controllers\\LegalsController@viewCreate");
            Route::get('/{id}', "\\Admin\\Controllers\\LegalsController@viewItem");
            Route::get('/{id}/delete', "\\Admin\\Controllers\\LegalsController@doDelete");
            Route::post('/create', "\\Admin\\Controllers\\LegalsController@doCreate");
            Route::post('/{id}/update', "\\Admin\\Controllers\\LegalsController@doUpdate");
        });
```

опять же вполне нормально переписывается в виде resource-роутов. за исключением отдельных методов (`cities.hide`, `cities.show`), которые скорее всего тоже либо реализуются в правильно описанном `update` или добавляются к resource-роутам:

```
Route::post('cities/{id}/hide', "\\Admin\\Controllers\\ContactsController@hide");
Route::post('cities/{id}/show', "\\Admin\\Controllers\\ContactsController@show");
// additional routes must prepend resource definition
Route::resource('cities', "\\Admin\\Controllers\\ContactsController");
```

#### https://bitbucket.org/rutorika/homekrasnodar_l5/src/2eeac9ca3907a426c5786a05118955376f406e73/app/Http/routes.php?at=master&fileviewer=file-view-default

``` php
Route::group(['middleware' => 'auth', 'namespace' => 'Dashboard', 'prefix' => 'dashboard'], function () {
    Route::get('/', ['uses' => 'MainController@index', 'as' => 'dashboard.main']);
    Route::resource('galleries', 'GalleryController');
});
```

это как раз пример использования resource-определения. но и тут есть ошибка в том, что используется множественное число `galleries`. нужно исользовать единственное `gallery`

#### https://bitbucket.org/rutorika/awi/src/8de4dc91ddf02cd936c67c9b5e4083b966a259e1/app/routes.php?at=master&fileviewer=file-view-default

``` php
// helpers.php
    function generate_crud_routes($routes, $entityDefaultNameSpace = '\\')
    {
        foreach ($routes as $route) {
            $name = $route['name'];
            $entity = camel_case($name);
            $controller = studly_case($name) . 'Controller';
            $prefix = array_key_exists('prefix', $route) ? $route['prefix'] : '';
            $entityClassName = studly_case($name);
            $entityNameSpace = isset($route['entityNameSpace']) ? $route['entityNameSpace'] : $entityDefaultNameSpace;
            Route::model($entity, $entityNameSpace . "{$entityClassName}");
            Route::get( "{$name}/{id}",                         ["as" => ".{$name}.view",      "uses" => "{$controller}@view"]);
            Route::get( "{$prefix}{$name}",                     ["as" => ".{$name}.index",     "uses" => "{$controller}@index"]);
            Route::get( "{$prefix}{$name}/create",              ["as" => ".{$name}.create",    "uses" => "{$controller}@create"]);
            Route::post("{$name}/store",                        ["as" => ".{$name}.store",     "uses" => "{$controller}@store"]);
            Route::post("{$name}/{id}/update",                  ["as" => ".{$name}.update",    "uses" => "{$controller}@store"]);
            Route::get( "{$name}/{id}/destroy",                 ["as" => ".{$name}.destroy",   "uses" => "{$controller}@destroy"]);
        }
    }
// routes.php
Route::group(
    ['before' => ['auth'], 'prefix' => 'dashboard', 'namespace' => 'App\Dashboard\Controllers'],
    function () {
        Route::get('/', ['as' => '.main', 'uses' => 'MainController@index']);

        $crudRoutes = [
            ['name' => 'city'],
            ['name' => 'placemark-type'],
            ['name' => 'building'],

            ['name' => 'action', 'prefix' => 'building/{building}/'],
            ['name' => 'placemark', 'prefix' => 'building/{building}/'],
            ['name' => 'liter', 'prefix' => 'building/{building}/'],
        ];

        generate_crud_routes($crudRoutes, 'App\\Entities\\');
    }
);
```

на самом деле вложенные роуты тоже поддерживаются в ларавел (http://laravel.com/docs/5.1/controllers#restful-nested-resources), поэтому и этот случай нормально представляется в виде ресурсов

``` php
Route::resource('building', 'BuildingController');
Route::resource('building.action', 'BuildingActionController');
```

Единственное, что `generate_crud_routes` ещё и генерирует имя контроллера по имени роута (`building` -> `BuildingController`)

**Comment by boxfrommars** [comment]
итак на повестке:
- [ ] описание роутов? если не берём стандартное, то описываем сами соответствия между действием и методом, и пишем свой аналог `Route::resource`, с нужными нами действиями
- [ ] используем ли ограничения параметров?
- [ ] описываем файл шаблон для роутинга, с правильной группировкой, миддлваре и всем таким и на его основе в следующих проектах делаем файлы `routes.php`

**Comment by boxfrommars** [comment]
Давайте сначала определимя с роутами, а потом перйдём к обсуждению контроллеров/crud-контроллеров и json api?

**Comment by demidovich** [comment]
Насколько я понимаю, нас интересуют роуты либо для админки либо для псевдо-rest. Я за стандартизированно-гибкое описание. То есть мы обозначаем структуру, как это делается и закладываем в нее возможность наглядно докидывать недостающие действия. Например, в проекте Compass у меня почти в каждом ресурсе было действие публикации чего-либо (тот самый publish). Вот он и будет описан отдельно. Но я за максимальную наглядность и понятность кода.

Так же считаю, что имеет смысл устанавливать глобальные проверки для параметров роутов. Описывать их один раз в начале роутера и все. То есть, если у нас есть id, то с вероятностью 100% это будет int или bigint. Таких параметров будет не много и это будет очень удобно.

**Comment by demidovich** [comment]
Следующим шагом предлагаю сделать по реализации каждому и выбрать из них лучшую.

**Comment by boxfrommars** [comment]
по идее публикацию/снятие с публикации можно решать и с помощью стандартных resource роутов, для этого есть [`PATCH` метод](http://jasonsirota.com/rest-partial-updates-use-post-put-or-patch), на который нужно посылать только изменившиеся поля:

```
PATCH /photo/{photo} 
published: true
```

и если сделать аккуратный crud, то такой запрос должен обрабатываться как 'изменить только пришедшие поля'.

но всё равно, конечно, кастомные дополнительные роуты должны быть.

**Comment by demidovich** [comment]
Согласен

**Comment by demidovich** [comment]
Хотя нет, не совсем согласен :) Несколько раз делал crud вместе с магическими видами, которые делали все. Это было удобно ровно до того момента, пока не прошло два месяца и не пришла задача - добавить новую сущность, но сделать в ней виды не укладывающиеся в то, что может сделать универсальный вид. В этом месте все становилось очень неудобно и очень стремно. Как это все в саппорте, я даже боюсь предположить.

В результате набитых шишек для себя я пришел к простой идеологии:
1. crud в контроллере с возможностью перезаписи вообще всего
2. роутинг (обсудили уже)
3. виды - тупой копипаст

Дак вот. В видах мне показалось красивее писать route('admin.entity.publish', $id). Именно так появился метод publish.

**Comment by boxfrommars** [comment]
> Это было удобно ровно до того момента, пока не прошло два месяца и не пришла задача - добавить новую сущность, но сделать в ней виды не укладывающиеся в то, что может сделать универсальный вид
> 
> crud в контроллере с возможностью перезаписи вообще всего
> виды - тупой копипаст

да, это как раз большие плюсы для resource, где каждому логическому действию соответствует отдельный метод в контроллер. согласен с ними. 

насчёт копипасты: можно избавиться от неё на уровне видов просто инклюдом общей формы/вида, до того момента как понадобится разделить поведение, например:

``` php
// show.blade.php
@include('_form.blade.php')

// edit.blade.php
@include('_form.blade.php')

// create.blade.php
@include('_form.blade.php')
```

и если вдруг понадобится новое поведение, просто добавляем `_edit_form.blade.php` или `_create_form_blade.php`

насчёт паблиш, сейчас попробую набросать оба варианта (отдельный метод vs `PATCH`), чтобы посмотреть и обсудить, что выглядит удобнее и понятнее

**Comment by boxfrommars** [comment]
**PATCH**
вид:

``` php
Form::open(['method' => 'PATCH', 'url' => route('article.update', $article->id)])
    Form::hidden('is_published', !$article->is_published);
    Form::submit($article->is_published ? 'Снять с публикации' : 'Опубликовать')
Form::close();
```

`routes.php`

``` php
Route::resource('article', 'ArticleController');
```

контроллер менять не нужно

**отдельный метод**
вид:

``` blade
@if ($article->is_published)
    <a href="{{ route('admin.entity.unpublish', $article->id) }}">Снять с публикации</a>
@else
    <a href="{{ route('admin.entity.publish', $article->id) }}">Опубликовать</a>
@endif
```

`routes.php`

``` php
Route::any('article/{id}/publish', ['as' => 'article.publish', 'uses' => 'ArticleController@publish'])
Route::any('article/{id}/unpublish', ['as' => 'article.unpublish', 'uses' => 'ArticleController@unpublish'])
Route::resource('article', 'ArticleController');
```

`ArticleController.php`

``` php
public function publish($id)
{
    $article = $this->_getEntity($id);
    $article->update(['is_published' => true]);
    return 'something';
}

public function unpublish($id)
{
    $article = $this->_getEntity($id);
    $article->update(['is_published' => false]);
    return 'something';
}
```

**Comment by boxfrommars** [comment]
у варианта с отдельным методом проще вид, у варианта с PATCH -- защита от csrf (хотя её можно и во втором варианте сделать, но тогда вид будет выглядеть также)

**Comment by demidovich** [comment]
Инклуды клево! Я вообще забыл про них в этом контексте :)

**Comment by demidovich** [comment]
publish применялся в таком контексте, я потому и вел речь про route()
http://joxi.ru/RmzXV6jCv4MnAO

**Comment by boxfrommars** [comment]
> Так же считаю, что имеет смысл устанавливать глобальные проверки для параметров роутов. Описывать их один раз в начале роутера и все. То есть, если у нас есть id, то с вероятностью 100% это будет int или bigint. Таких параметров будет не много и это будет очень удобно.

вот насчёт этого у меня сомнения. попробуй убрать эту проверку, изменится ли что-нибудь в поведении приложения?

ведь:

``` php
Article::find(7777777);
// null

Article::find('SomeBadString');
// null
```

никакой разницы, а проверка на существование article всё равно должна быть в коде, так что неважно, строка это или несуществующий в базе id. 
на мой взгляд, в случае с ограничением на id в роутах, мы просто добавляем лишние проверки, которые никак не изменяют поведение

единственный случай, когда это ограничение изменит поведение будет в том случае, если  у нас странный роутинг со странным порядком роутов, например:

``` php
Route::get('/article/{id}', "ArticleController@show");
Route::get('/article/something', "ArticleController@something");
```

в этом случае, без ограничений при запросе `/article/something` всегда будет обрабатываться первое правило, с ограничением -- второе. Но это решается простым правилом -- общие роуты должны быть ниже частных

**Comment by boxfrommars** [comment]
да, я тоже про этот же контекст. вот так, например, я делаю удаление в списках:

`index.blade.php`

```
@foreach($articles as $article)
    <tr>
        <td>{{ $article->id }}</td>
        <td>{{ delete_form(route('article.delete', $article->id)) }}</td>
    </tr>
@endforeach
```

`helpers.php`

``` php
function delete_form($url, $label = '<i class="glyphicon glyphicon-remove"></i>')
{
    $form = Form::open(['method' => 'DELETE', 'url' => $url]);
    $form .= '<button type="submit" class="btn btn-link">' . $label . '</button>';
    $form .= Form::close();
    return $form;
}
```

и получается:
![Grid with delete forms example](https://habrastorage.org/files/b2a/380/96b/b2a38096b6e648978a464430e1537673.png)

просто правильнее изменять и удалять с помощью `PUT`/`POST`/`PATCH`/`DELETE` (только для таких запросов работает csrf-защита), поэтому я и делаю с помощью форм

**Comment by boxfrommars** [comment]
так, думаю, нужно закрывать задачу. вот ответы на поставленные вопросы и задачи:

> описание роутов? если не берём стандартное, то описываем сами соответствия между действием и методом, и пишем свой аналог Route::resource, с нужными нами действиями

используемый стандартный ресурс-роутинг, дополняем необходимыми роутами

> используем ли ограничения параметров?

нет, не используем, т.к. всегда существуют дополнительные проверки (напр. на несуществование), которые невозможно реализовать с помощью ограничения параметров

> описываем файл шаблон для роутинга, с правильной группировкой, миддлваре и всем таким и на его основе в следующих проектах делаем файлы routes.php

https://gist.github.com/boxfrommars/d0416d871d5e39536cb5

обратите внимание на группировку, указывание префикса, неймспейса, документация по роутам: http://laravel.com/docs/5.1/routing

> Model binding

документация по биндингу моделей появилась: http://laravel.com/docs/5.1/routing#route-model-binding. думаю решение об использовании этого функционала должно приниматься самим разработчиком и не должно регулироваться правилами. скорее всего, после фиксации стандартных круд-контроллеров этот вопрос отпадёт сам собой

давайте до понедельника решим все возникшие дополнительные вопросы/замечания/пожелания и зафиксируем правила написания роутов
